sonic_y_cable/y_cable_base.py

```diff
@@ -0,0 +1,577 @@
+                 and their corresponding values
+
+        """
+
+        raise NotImplementedError
+
+    def download_firmware(self, port, fwfile):
+        """ This routine should download and store the firmware on all the
+        components of the Y cable of the port specified.
+        This should include any internal transfers, checksum validation etc.
+        from TOR to TOR or TOR to NIC side of the firmware specified by the fwfile.
+        This basically means that the firmware which is being downloaded should be
+        available to be activated (start being utilized by the cable) once this API is
+        successfully executed.
+        Note that this API should ideally not require any rollback even if it fails
+        as this should not interfere with the existing cable functionality because
+        this has not been activated yet.
+
+        Args:
+            physical_port:
+                 an Integer, the actual physical port connected to a Y cable
+            fwfile:
+                 a string, a path to the file which contains the firmware image.
+                 Note that the firmware file can be in the format of the vendor's
+                 choosing (binary, archive, etc.). But note that it should be one file
+                 which contains firmware for all components of the Y-cable
+        Returns:
+            an Integer:
+                 a predefined code stating whether the firmware download was successful
+                 or an error code as to what was the cause of firmware download failure
+        """
+
+        raise NotImplementedError
+
+    def activate_firmware(self, port, fwfile):
+        """ This routine should activate the downloaded firmware on all the
+        components of the Y cable of the port specified.
+        This API is meant to be used in conjunction with download_firmware API, and
+        should be called once download_firmware API is succesful.
+        This means that the firmware which has been downloaded should be
+        activated (start being utilized by the cable) once this API is
+        successfully executed.
+
+        Args:
+            physical_port:
+                 an Integer, the actual physical port connected to a Y cable
+        Returns:
+            an Integer:
+                 a predefined code stating whether the firmware activate was successful
+        """
+
+        raise NotImplementedError
+
+    def rollback_firmware(self, port):
+        """ This routine should rollback the firmware to the previous version
+        which was being used by the cable. This API is intended to be called when the
+        user either witnesses an activate_firmware API failure or sees issues with
+        newer firmware in regards to stable cable functioning.
+
+        Args:
+            physical_port:
+                 an Integer, the actual physical port connected to a Y cable
+        Returns:
+            an Integer:
```

Comment:
Maybe rather than explicitly stating "An Integer", we simply state "One of the following predefined constants:". This would also allow us to change the underlying type transparently if desired.

Also, please explicitly list the acceptable constants this function should return.

Reply:
fixed